cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/spark/ArtifactFetcherService.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.spark;
+
+import com.google.common.io.Files;
+import com.google.common.util.concurrent.AbstractIdleService;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.http.CommonNettyHttpServiceBuilder;
+import io.cdap.cdap.common.lang.jar.BundleJarUtil;
+import io.cdap.http.NettyHttpService;
+import org.apache.twill.filesystem.LocalLocationFactory;
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Launches an HTTP server for fetching artifacts.
+ */
+public class ArtifactFetcherService extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(ArtifactFetcherService.class);
+
+  private final NettyHttpService httpService;
+  private final Path workingDirectory;
+  private Location bundleLocation;
+
+  public ArtifactFetcherService(CConfiguration cConf, Path workingDirectory) {
```

Comment:
I think it would be a little cleaner if the bundle is created outside of this class and passed in. Then the bundleLocation can be final, the handler doesn't need a supplier, and we decouple building the jar from serving it.

Reply:
Agree. Fixed it.